pkg/crds/upgrade.go

```diff
@@ -0,0 +1,190 @@
+	if err != nil {
+		log.Fatal(err)
+		return nil, err
+	}
+
+	return &Upgrader{
+		client:    c,
+		namespace: namespace,
+	}, nil
+}
+
+// Upgrade installs the missing CRDs or updates them if they exists already
+func (u *Upgrader) Upgrade(targetVersion string) ([]unstructured.Unstructured, error) {
+	extractDir, err := helmutil.DownloadChartRelease(targetVersion)
+	if err != nil {
+		return nil, err
+	}
+
+	// reaper and medusa subdirs have the required yaml files
+	chartPath := filepath.Join(extractDir, helmutil.ChartName)
+	defer os.RemoveAll(chartPath)
+
+	crds := make([]unstructured.Unstructured, 0)
+
+	// For each dir under the charts subdir, check the "crds/"
+	paths, _ := findCRDDirs(chartPath)
+
+	for _, path := range paths {
+		err = parseChartCRDs(&crds, path)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for _, obj := range crds {
+		existingCrd := obj.DeepCopy()
+		log.Printf("Finding %s", obj.GetName())
+		err = u.client.Get(context.TODO(), client.ObjectKey{Name: obj.GetName()}, existingCrd)
+		if apierrors.IsNotFound(err) {
+			log.Printf("Creating %v\n", obj.GetName())
+			if err = u.client.Create(context.TODO(), &obj); err != nil {
+				log.Fatalf("Failed to create %s: %v\n", obj.GetName(), err)
+			}
+		} else if err == nil {
+			log.Printf("Updating %v\n", obj.GetName())
+			obj.SetResourceVersion(existingCrd.GetResourceVersion())
+			if err = u.client.Update(context.TODO(), &obj); err != nil {
+				log.Fatalf("Failed to update %s: %v\n", obj.GetName(), err)
+				return nil, err
+			}
+		} else {
+			log.Fatalf("Failed to Get the object %s: %v\n", obj.GetName(), err)
+			return nil, err
+		}
+	}
+
+	return crds, err
+}
+
+func findCRDDirs(chartDir string) ([]string, error) {
+	dirs := make([]string, 0)
+	err := filepath.Walk(chartDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			if strings.HasSuffix(path, "crds") {
+				dirs = append(dirs, path)
+			}
+			return nil
+		}
+		return nil
+	})
+	return dirs, err
+}
+
+func parseChartCRDs(crds *[]unstructured.Unstructured, crdDir string) error {
+	errOuter := filepath.Walk(crdDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		// Add to CRDs ..
+		b, err := ioutil.ReadFile(path)
+		if err != nil {
+			return err
+		}
+
+		dec := deser.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+
+		if len(b) == 0 {
+			log.Printf("Skipping %s\n", path)
+			// TODO Implement skipping this, it's a YAML with "---" starter
+			return nil
+		}
+
+		crd := unstructured.Unstructured{}
+		_, gvk, err := dec.Decode(b, nil, &crd)
+		if err != nil {
+			return err
+		}
+
+		if gvk.Kind != "CustomResourceDefinition" {
+			return nil
+		}
+
+		*crds = append(*crds, crd)
+		// }
+
+		// reader := k8syaml.NewYAMLReader(bufio.NewReader(bytes.NewReader(b)))
+
+		_ = k8syaml.NewYAMLOrJSONDecoder(bytes.NewReader(b), 4096)
```

Comment:
What is the purpose of this line?

Reply:
I still need to fix one thing, that is some CRDs could be built with "---" as starting and that messes up.